ow_plexil/src/plexil-adapter/LanderInterface.cpp

```diff
@@ -0,0 +1,349 @@
+                         "/TaskDeliverSample/status");
+    connectActionServer (m_panTiltMoveJointsClient, Name_PanTiltMoveJoints);
+    connectActionServer (m_cameraCaptureClient, Name_CameraCapture,
+                         "/CameraCapture/status");
+  }
+}
+
+
+/////////////////////////////// Unified Lander Interface ////////////////////////
+
+void LanderInterface::taskDeliverSample (int id)
+{
+  if (! markOperationRunning (Name_TaskDeliverSample, id)) return;
+  thread action_thread (&LanderInterface::nullaryAction<
+                        TaskDeliverSampleActionClient,
+                        TaskDeliverSampleGoal,
+                        TaskDeliverSampleResultConstPtr,
+                        TaskDeliverSampleFeedbackConstPtr>,
+                        this, id, Name_TaskDeliverSample,
+                        std::ref(m_taskDeliverSampleClient));
+  action_thread.detach();
+}
+
+
+void LanderInterface::armMoveCartesian (int frame,
+                                       bool relative,
+                                       const vector<double>& position,
+                                       const vector<double>& orientation,
+                                       int id)
+{
+  if (! markOperationRunning (Name_ArmMoveCartesian, id)) return;
+
+  geometry_msgs::Quaternion qm;
+
+  // Deal with type of orientation
+  if (orientation.size() == 3) { // assume Euler angle
+    tf2::Quaternion q;
+    // Yaw, pitch, roll, respectively.
+    q.setEuler (orientation[2], orientation[1], orientation[0]);
+    q.normalize();  // Recommended in ROS docs, not sure if needed here.
+    qm = tf2::toMsg(q);
+  }
+  else { // assume quaternion orientation
+    qm.x = orientation[0];
+    qm.y = orientation[1];
+    qm.z = orientation[2];
+    qm.w = orientation[3];
+  }
+
+  geometry_msgs::Pose pose;
+  pose.position.x = position[0];
+  pose.position.y = position[1];
+  pose.position.z = position[2];
+  pose.orientation = qm;
+  thread action_thread (&LanderInterface::armMoveCartesianAction, this,
+                        frame, relative, pose, id);
+  action_thread.detach();
+}
+
+void LanderInterface::armMoveCartesianAction (int frame,
+                                             bool relative,
+                                             const geometry_msgs::Pose& pose,
+                                             int id)
+{
+  ArmMoveCartesianGoal goal;
+  goal.frame = frame;
+  goal.relative = relative;
+  goal.pose = pose;
+  string opname = Name_ArmMoveCartesian;
+
+  runAction<actionlib::SimpleActionClient<ArmMoveCartesianAction>,
+            ArmMoveCartesianGoal,
+            ArmMoveCartesianResultConstPtr,
+            ArmMoveCartesianFeedbackConstPtr>
+    (opname, m_armMoveCartesianClient, goal, id,
+     default_action_active_cb (opname),
+     default_action_feedback_cb<ArmMoveCartesianFeedbackConstPtr> (opname),
+     default_action_done_cb<ArmMoveCartesianResultConstPtr> (opname));
```

Comment:
Do these 2 functions need to be separate?

Reply:
I don't see an obvious way to combine them since their argument types seem disjoint.